### DISCUSSION TITLE
JavaFX

### DISCUSSION THREAD

**Discussion by vsquared** [py5coding/py5generator#462] [Show and tell]
When trying to run the JavaFX code shown below, I get this error:
<img width="333" alt="jfx" src="https://github.com/py5coding/py5generator/assets/2294268/3de6dbcc-c715-4057-ba26-742984438998">

Do you know a way that I could add JavaFX to the Thonny IDE?  In Processing I can usually add all seven jar files to a data folder in the sketch folder and get it to run.

```
from javafx.scene import Scene
from javafx.stage import Stage
from javafx.scene.control import Button
from javafx.scene.layout import Pane

def setup():
    size(1,1,FX2D)
    stage = Stage()
    stage.set_title("Button demo")
    btn = Button("Button")
    btn.layout_x(80)
    btn.layout_y(80)
    pane = Pane()
    pane.get_children().add(btn)
    scene = Scene(pane, 250, 200)
    stage.setScene(scene)
    stage.show()
    
```

**Comment by hx2A** [comment]
Hi @vsquared !

You can put the javafx jars in a jars subdirectory. py5 will check for the existence of such a directory and will add those jars to the classpath before starting the JVM. Jars must always be added to the classpath before the JVM is started.

There is a good discussion about this that I will refer you to:

https://discourse.processing.org/t/importing-a-java-pocessing-library-into-py5/44444/8

This will provide you with more information on importing jars.

However, there is an extra twist: javafx has some weird import thing where you'll get a strange error when you try to add those jars to the classpath. And I vaguely remember getting py5 working with javafx at one point but stopped working on it. It can work but the weird error messages got annoying.

**Reply by vsquared** [reply]
I've seen the reference and I'm confused about the folder system.  I use _Import py5_ from the _py5_ Thonny menubar item for just about everything so far.  My resultant xxxx.py files are loose in a folder that I created for organizational purposes.  There is no folder system created when you use py5 this way.  I could create a folder for my xxxx.py file with a 'jars' subdirectory and put my javax jar files in there; is that what you have to do?

**Reply by hx2A** [reply]
If you are going to do a lot of stuff with jars, it might be convenient to make a jar subdirectory in the folder where you have the various code files. Then all of them will look in the same place for the jars and you won't need to copy them for each thing you are doing that creates them.

**Reply by vsquared** [reply]
Good idea.  Thanks.   Swing works and I'd like to get JavaFX working also.

**Reply by vsquared** [reply]
I created a subfolder called 'jars' inside of my 'Python' folder where I keep all of my xxxx.py files.  Inside of the 'jars' folder I added 22 files: the javafx jar files and associated lib files.  Then I ran the above demo code and the following error message was returned:
```
Traceback (most recent call last):
  File "Control.java", line 98, in javafx.scene.control.Control.<clinit>
java.lang.java.lang.IllegalStateException: java.lang.IllegalStateException: Toolkit not initialized

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "org.jpype.pkg.JPypePackage.java", line -1, in org.jpype.pkg.JPypePackage.getObject
java.lang.java.lang.ExceptionInInitializerError: java.lang.ExceptionInInitializerError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/tools/run_sketch.py", line 58, in <module>
    main()
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/tools/run_sketch.py", line 49, in main
    imported.run_code(
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 148, in run_code
    _run_code(
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 304, in _run_code
    _run_sketch(sketch_path, classpath, exit_if_error)
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/py5_tools/imported.py", line 295, in _run_sketch
    exec(sketch_compiled, py5_ns)
  File "/Users/xxxx/Python/py5_javafxDemo.py", line 3, in <module>
    from javafx.scene.control import Button
  File "/Users/xxxx/Library/Python/3.10/lib/python/site-packages/jpype/imports.py", line 82, in _JExceptionHandler
    raise ImportError("Unable to import '%s' due to initializer error" % javaname) from ex
ImportError: Unable to import 'javafx.scene.control.Button' due to initializer error
```

We're getting closer, but not quite there.

**Reply by hx2A** [reply]
Right, I vaguely remember that error. There's some other Java code that needs to execute to enable JavaFX. I remember researching this a long time ago in the Processing source code. I would suggest looking there and see what it does to get JavaFX working in a Sketch. I know that's a non-trivial task but that's where the answer is going to be.

**Reply by vsquared** [reply]
Here's a bare bones Processing JavaFX demo:
```
import processing.javafx.*;
import javafx.stage.Stage;
import javafx.scene.Scene;
import javafx.scene.layout.Pane;

void setup(){
  size(1,1, FX2D);
  Stage stage = new Stage();
  stage.setTitle("JavaFX Demo");
  Pane pane = new Pane(); 
  Scene scene = new Scene(pane, 400, 400);
  stage.setScene(scene);
  stage.show();  
}

void draw(){
}
```
It needs a _processing.javafx.*_ file in addition to the ones that I listed.  I have no idea where to get that file and where to put it.

**Reply by hx2A** [reply]
> It needs a processing.javafx.* file in addition to the ones that I listed. I have no idea where to get that file and where to put it.

Everything you need is in here:

https://github.com/processing/processing4-javafx/releases/download/latest/processing4-javafx.zip

I would download that and unzip it and put it in your jars subdirectory.

The javafx.jar file contains the processing.javafx.* stuff (I believe). The other platform specific folders contain native libraries and more JavaFX jars that are separate from Processing. Those Jars are "modules" and need to be added to the JVM but it wants to add them in a way that is different from the way they are normally added to the classpath. That's the weird error I was seeing.

I just reviewed my code from when I last looked at this in 2022 and I see that I did in fact get it to work. The following code works on my old MacOS computer right now:

```python
import py5_tools

py5_tools.add_options('-Djava.library.path=jars/javafx/library/macos-x86_64/')
py5_tools.add_options('--module-path=jars/javafx/library/macos-x86_64/modules/')
py5_tools.add_options('--add-modules=ALL-MODULE-PATH')
py5_tools.add_options('--add-exports=javafx.graphics/com.sun.javafx.geom=ALL-UNNAMED')
py5_tools.add_options('--add-exports=javafx.graphics/com.sun.glass.ui=ALL-UNNAMED')

import py5


def setup():
    py5.size(300, 300, py5.FX2D)


def draw():
    py5.rect(py5.mouse_x, py5.mouse_y, 10, 10)


py5.run_sketch()
```

You'd need to change the platform specific stuff for your Mac computer's architecture. I have no idea if it will work on your machine but I know figuring out those options is non-trivial so this will be a big step forward with your goals.

Now I am curious why I stopped working on this. Maybe it works on this MacOS machine but there are other problems with Linux or Windows? I don't remember. There must have been a problem that made me pause this. In any case, I'll revisit this later in the week when I am back from PyCon and have more time.

**Reply by vsquared** [reply]
Thanks for the valuable tip.  In spite of the error shown below, it does run with a little re-arranging necessitated by other error messages (I realize that I called py5.rect() twice).  It is not what I expected but it does run and that's a giant leap for mankind.  It may just be ignoring all the javafx stuff and instead running the last few lines because we know that is good code.  At any rate, it didn't balk at the modules code at the top.  So now I need to see if I can actually use those modules.  The graphics module should allow text, but I couldn't get it to accept that.  I'll keep working with it.

<img width="718" alt="fx" src="https://github.com/py5coding/py5generator/assets/2294268/523c3ef4-258c-48d8-940d-716f54bff5d2">

**Error message:**
Unable to obtain source code for setup(). Either make it obtainable or create a settings() function for calls to size(), fullscreen(), etc.

I greatly appreciate your help with this.

**Reply by hx2A** [reply]
> Unable to obtain source code for setup(). Either make it obtainable or create a settings() function for calls to size(), fullscreen(), etc.

This error is unrelated to FX2D. py5 does some magic to enable you to put settings stuff in setup but for that to work it needs access to the source code, and for that to happen, the code must be saved to a file. If you put that code in a file and save it and run it from the command line with `python test.py` it will work. I agree that error message could be improved a bit and there is an open issue to address that.

>  It is not what I expected but it does run and that's a giant leap for mankind.

Indeed, it is!

Those calls to py5_tools must be made before the JVM starts which means that using Thonny with imported mode will not work here. I recall I had the idea that py5 would detect if the user is attempting to use FX2D and would run that stuff before starting the JVM. That is not that difficult to do but I don't think that is the only challenge of using FX2D.

You're now at the limit of what I figured out with FX2D. Please keep me in the loop as you progress on this front.

> I greatly appreciate your help with this.

You are welcome!

**Reply by vsquared** [reply]
`If you put that code in a file and save it and run it from the command line with python test.py it will work. ` 
Confirmed.  On a mac I have to use _python3_.

**Reply by vsquared** [reply]
I'm trying to use those wonderful javafx controls and I keep getting error messages.  eg I added _scene = Scene()_ to your code and I get `NameError: name 'Scene' is not defined`.

**Reply by hx2A** [reply]
What is Scene? A Java class? I suggest looking at py5's documentation on hybrid programming:

https://py5coding.org/content/hybrid_programming.html

**Reply by vsquared** [reply]
Look back at the first post.  Scene is like the window.  We also need Pane and all the controls.  In my opinion what you have done is to make _FX2D_ available.  What you have not done is make all the modules available.  You correctly tried to set the modulePath and load all the modules, but I'm not sure that's working.  Currently all that your javafx model can do is put up a primitive graphics, eg rect and circle.  I'm unable to use it for text using your sample code on the webpage and it won't recognize the controls.

**Reply by hx2A** [reply]
Please read the hybrid coding documentation. In the hybrid coding documentation, it explains how to import Java classes and make them available to Python via JPype with `JClass`.

You may need to make more calls to `add_options()` with `--add-exports` parameters, depending on what you are trying to do, but this is something you will need to figure out from looking Java FX documentation.

**Reply by vsquared** [reply]
I've built lots of JavaFX demos using Processing and was working on a JavaFX Editor in objc before one of your contributors convinced me to get involved with Python, so I have a pretty good idea how to use it (including from the command line with javac and java).  Porting the code from Processing to py5 is something else.  Swing components ported fairly easily, but I'm not too sure about JavaFX.  What I'm trying to do is straight forward, I just need access to all seven of the modules.  The editors of Processing for some reason won't add all seven to their editor; they did include two or three modules, but not all of them.  There is one line of code that needs about three or four more modules added to it.  I even told them what line of code needed to be changed and there was one response that made no sense to me.  They already included all the modules in the _javafx_ folder that goes in the 'libraries' folder; all that they have to do is amend the code to access them.  I modified their code and recompiled the editor myself, therefore in my copy of the Processing editor I can use any JavaFX control that I want.  The problem is no one else can run my demos, because they're stuck with the three modules that they were given (some users are savvy enough to add all seven modules to a 'data' folder in the sketch folder and get by).  It looks like the Processing editors use some JavaFX controls in their work, but won't support the rest of us using them.  If we get py5 to use all of JavaFX it will be a first for this community.

**Reply by hx2A** [reply]
> I  just need access to all seven of the modules.

I understand. Sounds great. I believe you will be able to add the other modules by writing more `py5_tools.add_options()` code. Is that not the case?

**Reply by vsquared** [reply]
In your demo you added all the modules.  I'd like to try another approach while I try to understand how py5 and pype are constructed.  I may disappear for a couple of days because of family matters, but I will keep working on it in my free time.  The approach that I would like to use for openers is the same one that I used for Swing, ie take proven Processing code and try to convert it to Python using 'Import mode for py5'.  Just because we placed the seven jar files in the 'jars' folder does not mean that py5 knows how to access them and that seems to be where it's hung up.  Those files are not being initialized and Thonny is telling me that the hang up is in JPype.  Below is the code that I am trying to run and I may shorten it even more to a single jar file.  Following that are a series of screenshots pointing to where Thonny says the problem lies.  You should be able to confirm with your system.  I realize that it is a different approach than your demo with all the py5. prefixes, but I want to try it.

**Source code:**
```
import javafx.stage.Stage;
import javafx.scene.Scene;
import javafx.scene.layout.Pane;
import javafx.scene.control.Button;

def setup():  
    size(1, 1, FX2D)
    stage = Stage()
    stage.setTitle("Button demo")
    btn = Button("button")
    btn.setLayoutX(80)
    btn.setLayoutY(80)
    pane = Pane()
    pane.getChildren().add(btn)
    scene = Scene(pane, 250, 200)
    stage.setScene(scene)
    stage.show()  

```
**Errors:**
<img width="753" alt="shell err" src="https://github.com/py5coding/py5generator/assets/2294268/7009ad5c-c6b9-45f2-8c53-2280ce9065ae">

**Source of error is in here:**


<img width="657" alt="jpype" src="https://github.com/py5coding/py5generator/assets/2294268/678da7b5-f9f8-4317-9822-c2ed86d5506b">

**At this line:**
<img width="839" alt="pype err" src="https://github.com/py5coding/py5generator/assets/2294268/d86ab4de-aeaa-420f-8d7c-dcddccbdf147">

**Reply by hx2A** [reply]
> Those files are not being initialized and Thonny is telling me that the hang up is in JPype.

The files are not being initialized because the relevant code to initialize them was not used. In the example code I provided, there were several lines of code using py5_tools that enable the modules. You cannot use that code with imported mode, however, because that code can only be run before the JVM is started, which takes place when you import py5. In imported mode, this is done before the first line of your code executes, so you do not get a chance to run the necessary code needed to enable the FX modules. Therefore, FX2D is not compatible with imported mode.

> I realize that it is a different approach than your demo with all the py5. prefixes, but I want to try it.

If you are going to continue down this path of exploration, I am going to have to insist that you use module mode and stop trying to use imported mode. You should also know that I am also a person with needs outside of GitHub and am currently managing my Dad's health problems, which is consuming much of my time, and has been consuming my time for many months, and will continue to do so for the foreseeable future. Every day is a struggle as I try to manage my own needs on top of what I am doing for my parents. One of the things I do outside of that is try to grow the py5 library and community, and part of how I do that is by answering questions here. I am motivated for people to have a good experience with py5. I don't have a lot of time to do this. I am getting frustrated here because I am trying to help you with what you want to accomplish by pointing you in a direction that could work but you want to do things in a way that cannot possibly work. Asking me to help you do something that cannot work is not fair to me or my time.

**Reply by villares** [reply]
Cheers @vsquared and @hx2A !

vsquared, you  asked the other day the benefits of module mode, comparing with imported mode that takes less typing, and it was not that clear to me (other than namespace clashes) so I guess now we found a very clear limitation of imported mode and benefit of moving advanced stuff to module mode! 

Maybe we could add that to the docs, if you want to import jars, if you need to add stuff to jpype's classpath, you have to use module mode or class mode.

**Reply by vsquared** [reply]
`Asking me to help you do something that cannot work is not fair to me or my time.`

Please don't stress out about this.  It's not a life or death matter.  I have lots of projects I can work on and I'm not totally committed to py5.  If we get JavaFX to work in py5, great, if we don't it's no big deal.  I only report things to you because you said you wanted to be kept in the loop;  I certainly don't expect you to answer questions that you don't know the answers to, but I am always interested in what you have to say.  I have another opinion from someone on the JPype forum and he thinks that there's some problem with the JavaFX module jars and associated code and admits that there is a scarcity of error messages to solve the problem, but supposedly he has had JavaFX running in some other python app, so it appears to be possible.  For now I nearly have an educational app written in py5 with Swing components that should help students master the multiplication tables.  I'll post it here when I get it done and run it by Villares since I believe he is a teacher.  At least we were successful getting Swing into py5.  As far as JavaFX goes I also got it into Processing, but I don't see that many people using it.  It can also be run from the command line and I need to finish my objc based JavaFX editor so I can experiment more with it.  Creating a py5 editor that could run JavaFX from the command line is another possibility if this other technique proves too difficult to do.  At any rate, don't fret over it and don't think you have to provide answers.  I sympathize with you trying to care for an infirmed parent and realize it's not easy.

**Reply by hx2A** [reply]
@vsquared , I do want you to know I do respect your motivation to experiment and try out ideas, especially since the motivation to hack things is literally the foundation upon which py5 is born. I hope my tone does not come across in a way that diminishes your motivation. It's just that imported mode as a concept is suited for beginner coders and combining Java and Python is an advanced technique, and there will certainly be shortcomings in what something designed for beginners is able to achieve when it is used for advanced things. Those shortcomings are perhaps weaknesses but are also necessary parts of making something easier for beginners to use.

**Reply by hx2A** [reply]
(I typed my previous message before your reply appeared here)

> At any rate, don't fret over it and don't think you have to provide answers. I sympathize with you trying to care for an infirmed parent and realize it's not easy.

Thank you for being understanding, and you have a point, I am putting pressure on myself to provide answers. Maybe accepting that I don't have to do that all the time is something I need to learn as a maintainer. I appreciate your helping me realize that; I'm sure that will help me in the future.

**Comment by vsquared** [comment]
Success!  Should work for openers(seven javafx modules should be dropped into 'jars' folder):
```
import py5
import javafx.scene.Scene
import javafx.scene.layout.Pane
import javafx.stage.Stage

this = py5._instance

def settings():
    this.size(1,1,this.FX2D)
    
def setup():
    btn = javafx.scene.control.Button("Button")
    print(btn)   
# **** Adds btn to pane ****    
    pane = javafx.scene.layout.Pane(btn)
    print("pane =",pane) 
    scene = javafx.scene.Scene(pane,250,200)
    stage = javafx.stage.Stage()
    stage.setTitle("JavaFX Demo")
    print("stage =",stage)   
    print("scene = ",scene)  
    stage.setScene(scene)
    stage.show()
    
__name__ == '__main__' and py5.run_sketch()

```
**Output:**

<img width="254" alt="poc" src="https://github.com/py5coding/py5generator/assets/2294268/cff4bd6c-b153-414d-9eda-b35701e1fb28">

**Reply by villares** [reply]
This looks very cool and promising!

I can imagine a GUI library for py5 based on this, if you could package it (hide some of the complexity) in a way that would make it easy for people to both:
- Have a normal py5 canvas window and a "controls" window with JavaFX elements
- Have a JavaFX window like this one with an interactive py5 canvas inside.

**Reply by vsquared** [reply]
Thanks for reviewing.  I'm curious if you can run this in Linux and what does the output look like?  BTW the graphics and text with the FX2D renderer are really sharp compared to what we're used to in Processing.  Below is a window arrangement like what you described.
**FX2D graphics output:**
<img width="406" alt="py5_javafx_1" src="https://github.com/py5coding/py5generator/assets/2294268/4466d2f0-7b24-4d33-bc96-7665da70f10f">

**Reply by villares** [reply]
Oh boy, I was unlucky... I copied the jars but it doesn't work for me on Linux...

```
Graphics Device initialization failed for :  es2, sw
Error initializing QuantumRenderer: no suitable pipeline found
...
    import javafx.scene.layout.Pane
  File "/home/villares/.local/lib/python3.10/site-packages/jpype/imports.py", line 189, in find_spec
    if not hasattr(base, parts[2]):
  File "/home/villares/.local/lib/python3.10/site-packages/jpype/imports.py", line 82, in _JExceptionHandler
    raise ImportError("Unable to import '%s' due to initializer error" % javaname) from ex
ImportError: Unable to import 'javafx.scene.layout.Pane' due to initializer error
```

**Reply by hx2A** [reply]
I get the same exception as @villares on Linux. Glad you are making progress with this, @vsquared ! How does this work without enabling the FX modules? Is that somehow being done elsewhere?

**Comment by vsquared** [comment]
@villares 
Here's an example of building two windows in a single py5 file.  One window is the py5 window with FX2D renderer and the other is a JavaFX controller window.  The size of the circle in the py5 window may be changed by adjusting the slider in the JavaFX window.

```
import py5
import javafx.scene.Scene
import javafx.scene.layout.Pane
import javafx.stage.Stage
import javafx.scene
import javafx.scene.text.Text
import javafx.scene.text.Font
import javafx.scene.paint.Color

this = py5._instance

_wndW = 400
_wndH = 400

def settings():
    this.size(_wndW,_wndH,this.FX2D)
 
def setup():
    global stage
    global scene
    global pane
    global slider
    
    py5.get_surface().set_title("FX2D Window")

    slider = javafx.scene.control.Slider(0,300,100)
    print(slider)
    print("slider init value =",slider.getValue())  
    label = javafx.scene.control.Label("Pull slider to change circle.")
    print(label)
    btn = javafx.scene.control.Button("Button")
    print(btn)
    txt = javafx.scene.text.Text("Welcome to JavaFX")
    txt.setUnderline(True)
    print("txt =",txt)
    print(txt.getText())

    pane = javafx.scene.layout.Pane()
    print("pane =",pane)
    pane.getChildren().add(txt)
    txt.relocate(40,20)
    pane.getChildren().add(label)
    label.relocate(40,50)
    pane.getChildren().add(slider)
    slider.relocate(30,80)
    pane.getChildren().add(btn)
    btn.relocate(80,120)
    myColor = javafx.scene.paint.Color
    print("ALICEBLUE =",myColor.ALICEBLUE)   
    scene = javafx.scene.Scene(pane,250,200, myColor.ALICEBLUE)
    stage = javafx.stage.Stage()
    print("stage =",stage)
    stage.setTitle("JavaFX Window")
    print("scene = ",scene)  
    stage.setScene(scene)
    stage.show()

def draw():
    this.background(209)
    this.fill(255)
    this.rect(10,10,220,40)
    this.fill(0)
    this.textSize(22)
    this.text("Welcome to FX2D!!",20,40)
    this.fill(0,0,255)
    this.circle(_wndW/2,_wndH/2,float(slider.getValue())) 
    
py5.run_sketch()

```
**Output:**
<img width="677" alt="twoWnd" src="https://github.com/py5coding/py5generator/assets/2294268/ccbf7a7b-c245-4639-81c2-16dea8f41fd0">

**Reply by villares** [reply]
Well, I fiddled a bit more... downloaded some other JavaFX jars from https://gluonhq.com/products/javafx/ combined them with the javafx.jar from Processing... and...
![peek_1](https://github.com/py5coding/py5generator/assets/3694604/24801683-898f-47bc-8bf8-87d8d19ef1dd)

**Reply by vsquared** [reply]
I have a lot of jar files in my 'jars' folder and I'm still trying to determine what is the bare minimum.  Based on my work with the cmd-line you need at least javafx.controls.jar and perhaps javafx.graphics.jar depending on what you are trying to do.  I'm not sure if you need javafx.jar as well (I don't believe I ever called it from the cmd-line, but it has been awhile since I did that.)  When it comes to the 'jars' folder I probably have far too many files in there, but I always thought it was better to have too many than not enough.  Glad you got it working on Linux.
<img width="1474" alt="jars" src="https://github.com/py5coding/py5generator/assets/2294268/8b6d04fb-5a49-4b58-919f-ed23650e961f">

**Reply by hx2A** [reply]
@vsquared , can you show us the output of `tree jars/`? I can't read the filenames in that image, it is too blurry.

**Reply by villares** [reply]
@hx2A I haven't tried the last imported mode demo by @vsquared , but this is from the slider demo, using a jars from https://gluonhq.com/products/javafx/ -> ` javafx-sdk-22.0.1` (2024 dates) plus `javafx.jar` (2022, I think) from the Processing 4 link you sent us. I'm not 100% sure because I was in crazy-tired-eploratory mode, I might be able to try and replicate it taking notes at some later time...  

```
>>> for p in jpype.getClassPath().split(':'):
    print(p)
    
/home/villares/.local/lib/python3.10/site-packages/py5/jars/core.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/gluegen-rt.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/jogl-all.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/py5.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/dxf/dxf.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/pdf/itext.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/pdf/pdf.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/svg/batik.jar
/home/villares/.local/lib/python3.10/site-packages/py5/jars/svg/svg.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.base.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.controls.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.fxml.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.graphics.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.media.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.swing.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx.web.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/javafx-swt.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.base.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.controls.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.fxml.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.graphics.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.media.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.swing.jar
/home/villares/GitHub/python_play_reboot/py5_play/jars/linux-amd64/modules/javafx.web.jar
```

**Comment by vsquared** [comment]
For those who believe that it's impossible to run JavaFX from Imported Mode:
```
# Uses Imported mode for py5

import javafx.scene.Scene
import javafx.scene.layout.Pane
import javafx.stage.Stage
import javafx.scene
import javafx.scene.text.Text
import javafx.scene.text.Font
import javafx.scene.paint.Color

def settings():
    size(1,1,FX2D)
 
def setup():    
    btn = javafx.scene.control.Button("Button")
    print(btn)
    txt = javafx.scene.text.Text("Welcome to JavaFX")
    txt.setUnderline(True)
    font = javafx.scene.text.Font("Verdana",25)
    txt.setFont(font)
    print("txt =",txt)
    print(txt.getText())
    pane = javafx.scene.layout.Pane()
    print("pane =",pane)
    pane.getChildren().add(txt)
    txt.relocate(40,20)
    pane.getChildren().add(btn)
    btn.relocate(80,120)
    myColor = javafx.scene.paint.Color
    print("ALICEBLUE =",myColor.ALICEBLUE)   
    scene = javafx.scene.Scene(pane,320,200, myColor.ALICEBLUE)
    stage = javafx.stage.Stage()
    print("stage =",stage)
    stage.setTitle("JavaFX Window")
    print("scene = ",scene)  
    stage.setScene(scene)
    stage.show()
    
run_sketch()
```
**Output:**
<img width="323" alt="importedMode" src="https://github.com/py5coding/py5generator/assets/2294268/a1b48e0f-2b5d-4bf5-aa84-e89005d3fb3a">

**Issue:**
Note that only the window title bar is ALICEBLUE (change to myColor.RED to make more obvious), whereas the pane that the buttons appear on is supposed to be colored, not the title bar.

**Reply by hx2A** [reply]
Interesting. Glad your persistence got results. It doesn't work on my machine though, and I'm confused how this can work. You are doing this on a mac, right? How exactly are you running this? Through thonny? Can it run from the command line with `run_sketch fx2d_imported_mode.py`? I'd like to see what is in your jars folder with `tree jars/`, if you could.

**Reply by vsquared** [reply]
There's a screen shot of my jar folder a couple of posts up.  Won't run from cmd-line with python3 fileName which is what I usually use.  May need to use module path addmodules ; I'll work on it.  It runs on macos Sonoma 14.4.1 with Thonny editor using py5 plugin and _Imported mode for py5_ checked on menubar.

**Addendum:**
run_sketch ... returns command not found

python3 fileName returns ModuleNotFoundError: No module named 'javafx'

**Reply by hx2A** [reply]
> There's a screen shot of my jar folder a couple of posts up.

GitHub was giving me problems before and wouldn't show me that full-sized. Now I can see it. Can you tell me what is in the javafx folder? What is the contents of javafx.properties?

> run_sketch ... returns command not found

Thonny uses the `run_sketch` command line utility that is installed along with py5. For `run_sketch` to work, it must be run in the same environment that Thonny is using.

> python3 fileName returns ModuleNotFoundError: No module named 'javafx'

If this is also run in the same place that you get `run_sketch ... returns command not found`, it will also fail because py5 is not there to activate jpype's ability to import Java classes.

> It runs on macos Sonoma 14.4.1 with Thonny editor using py5 plugin and Imported mode for py5 checked on menubar.

Since Thonny uses `run_sketch` when you check Imported mode for py5, is Thonny doing something magical on your machine to enable JavaFX? What version of py5 is installed there?

**Reply by vsquared** [reply]
javafx folder is found in the Processing/libraries folder and you also sent me a link to it.  I don't think it is used.  At some point I'm going to start stripping out files from the 'jars' folder until I find the bare minimum.  At this point my current concept is that we just need the seven jar modules. I'm not sure about all those lib files.

My concept: py5 is automatically called when I check Imported mode for py5 so I don't have to use the py5. prefix(Yay!!)  I don't think Thonny is a magician other than using the py5 plugin.  As mentioned previously, py5 version is the latest (I forget the numbers).

**Comment by vsquared** [comment]
Here's a revision of my original post which now runs without error  in the Thonny editor with py5 plugin and _Imported mode for py5_ checked in the menubar py5 menu.

```
import javafx.scene.Scene
import javafx.scene.layout.Pane
import javafx.stage.Stage

def settings():
    size(1,1,FX2D)
    
def setup():
    btn = javafx.scene.control.Button("Button")    
    btn.setLayoutX(80)
    btn.setLayoutY(80)
    pane = javafx.scene.layout.Pane()
    pane.getChildren().add(btn)
    scene = javafx.scene.Scene(pane, 250, 200)
    stage = javafx.stage.Stage()
    stage.setTitle("Button demo")
    stage.setScene(scene)
    stage.show()
```

**Reply by villares** [reply]
I have a `jars` folder next to the sketch's `javafx3.py` file. Some magic going on I suppose.
![image](https://github.com/py5coding/py5generator/assets/3694604/6b1d4512-8d76-40ca-9e1a-449e033194a2)

**Reply by hx2A** [reply]
@villares , can I see all of your classpath? Are the files in `linux-amd64/modules/` identical to the files in the directory two levels up? You are doing this on linux, right?

**Reply by vsquared** [reply]
For whatever it's worth:  I have a folder called "Python" in my home directory which contains almost every python file I ever created using Thonny (many made before I ever used py5).  Inside the "Python" folder is another folder called 'jars'; inside the  'jars' folder are the seven javafx jar files and other lib files as shown in the screenshot above.  Py5 code apparently is what is accessing jar files inside the 'jars' folder.

**Reply by villares** [reply]
Does this help, @hx2A ? I'm on Linux. They look pretty similar... I should have documented better... but as I said... crazy-tired-exploratory mode...

![image](https://github.com/py5coding/py5generator/assets/3694604/e4f23730-2bd9-4b4f-9042-084bed92f331)
![image](https://github.com/py5coding/py5generator/assets/3694604/12ea1415-ffa9-43c0-bb3e-af129dacbec1)
![image](https://github.com/py5coding/py5generator/assets/3694604/14450750-e41c-4f18-8215-dab118b00b83)

**Reply by vsquared** [reply]
@villares 
Thanks for confirming.

**Reply by vsquared** [reply]
@hx2A 
As far as I can tell the image below is a minimum of what is required in the jars folder; javafx.jar appears to be necessary (last demo broke when I removed it) in addition to the eight javafx.xxxx.jar files.  I'm still not sure about those lib files and have not experimented with removing them:

<img width="1334" alt="jars_fldr" src="https://github.com/py5coding/py5generator/assets/2294268/d90e6204-a3b4-47b7-8fbd-50fc186e7e09">
